xbuild/src/cargo/utils.rs

```diff
@@ -19,77 +19,132 @@ pub fn list_rust_files(dir: &Path) -> Result<Vec<String>> {
     let path = dunce::canonicalize(path)?;
-    for manifest_path in path
+    let manifest_path = path
         .ancestors()
         .map(|dir| dir.join("Cargo.toml"))
-        .filter(|dir| dir.exists())
-    {
-        let manifest = Manifest::parse_from_toml(&manifest_path)?;
-        if let Some(p) = manifest.package.as_ref() {
-            if let (Some(n1), n2) = (name, &p.name) {
-                if n1 == n2 {
-                    return Ok((manifest_path, p.name.clone()));
-                }
+        .find(|dir| dir.exists())
+        .ok_or(anyhow::anyhow!("Didn't find Cargo.toml."))?;
+
+    let manifest = Manifest::parse_from_toml(&manifest_path)?;
+
+    // This function shouldn't be called when a workspace exists.
+    anyhow::ensure!(
+        manifest.workspace.is_none(),
+        "Did not expect a `[workspace]` at `{}`",
+        manifest_path.display(),
+    );
+
+    if let Some(package) = &manifest.package {
+        if let Some(name) = name {
+            if package.name == name {
+                Ok((manifest_path, manifest))
             } else {
-                return Ok((manifest_path, p.name.clone()));
-            }
-        }
-        if let (Some(w), Some(name)) = (manifest.workspace.as_ref(), name) {
-            if let Some(manifest_path) = member(&manifest_path, &w.members, name)? {
-                return Ok((manifest_path, name.to_string()));
+                Err(anyhow::anyhow!(
+                    "package `{}` not found in workspace `{}`",
+                    manifest_path.display(),
+                    name,
+                ))
             }
+        } else {
+            Ok((manifest_path, manifest))
         }
+    } else {
+        Err(anyhow::anyhow!(
+            "Failed to parse manifest at `{}`: virtual manifests must be configured with `[workspace]`",
+            manifest_path.display(),
+        ))
     }
-    anyhow::bail!("cargo manifest not found");
 }
 
-pub fn find_workspace(manifest: &Path, name: &str) -> Result<Option<PathBuf>> {
-    let dir = manifest.parent().unwrap();
-    for manifest_path in dir
+/// Find the first `Cargo.toml` that contains a `[workspace]`
+pub fn find_workspace(potential_root: &Path) -> Result<Option<(PathBuf, Manifest)>> {
+    for manifest_path in potential_root
         .ancestors()
         .map(|dir| dir.join("Cargo.toml"))
-        .filter(|dir| dir.exists())
+        .filter(|manifest| manifest.exists())
     {
         let manifest = Manifest::parse_from_toml(&manifest_path)?;
-        if let Some(w) = manifest.workspace.as_ref() {
-            if member(&manifest_path, &w.members, name)?.is_some() {
-                return Ok(Some(manifest_path));
-            }
+        if manifest.workspace.is_some() {
+            return Ok(Some((manifest_path, manifest)));
         }
     }
     Ok(None)
 }
 
-/// Search for .cargo/config.toml file relative to the workspace root path.
-pub fn find_cargo_config(path: &Path) -> Result<Option<PathBuf>> {
```

Comment:
Moved to `impl Config`.

Reply:
why? does this make sense later? don't see it yet

Reply:
It cleans up a bit for when we start using `LocalizedConfig` (which stores the location of `.cargo/config.toml` so that it can later yield `[env]` vars with `relative = true` to the proper directory.